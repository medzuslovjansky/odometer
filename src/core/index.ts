export {
  Multireplacer,
  MultireplacerRule,
  ReplacementValue,
  ReplacementFunction,
} from './multireplacer';

export { Odometer } from './odometer/Odometer';

export { Intermediate, Splitters } from './utils';
