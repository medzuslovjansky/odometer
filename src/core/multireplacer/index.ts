export * from './Multireplacer';
export * from './MultireplacerRule';
export * from './Replacement';
