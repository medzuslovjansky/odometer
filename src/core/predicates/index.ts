export * from './FlavorizationPredicate';
export * from './FlavorizationPredicateFunction';
export * from './FlavorizationPredicateGroup';
export * from './FlavorizationPredicateObject';
export * from './FlavorizationPredicateWrapper';
