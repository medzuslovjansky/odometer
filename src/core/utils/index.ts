import * as Splitters from './Splitters';

export { Splitters };
export * from './Intermediate';
