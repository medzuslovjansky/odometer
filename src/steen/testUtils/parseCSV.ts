import _ from 'lodash';
import parse from 'csv-parse';

export async function parseCSV(
  buffer: Buffer | string,
): Promise<Record<string, string>[]> {
  const rawRecords = await new Promise<Record<string, string>[]>(
    (resolve, reject) => {
      parse(buffer, function (err, data) {
        if (err) {
          return reject(err);
        }

        const [headers, ...rows] = data as string[][];
        const records = rows.map((row) => {
          return _.transform(
            row,
            (acc, value, index) => {
              acc[headers[index]] = value;
            },
            {} as Record<string, string>,
          );
        });

        resolve(records);
      });
    },
  );

  return rawRecords;
}
