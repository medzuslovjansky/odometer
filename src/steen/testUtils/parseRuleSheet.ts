import { FlavorizationLevel } from '../FlavorizationLevel';
import { Multireplacer } from '../../core/rules/Multireplacer';
import { FlavorizationRuleDTO } from '../FlavorizationRuleDTO';
import { MultireplacerRule } from '../../core/rules';
import { parse as steenparse } from '@interslavic/steen-utils';
import { FlavorizationPredicate } from '../../core/predicates';
import { InterslavicIntermediate } from '../InterslavicIntermediate';
import _ from 'lodash';
import createMapReplacer from '../utils/createMapReplacer';
import { parseCSV } from './parseCSV';

export async function parseRuleSheet(
  buffer: Buffer | string,
): Promise<Record<keyof typeof FlavorizationLevel, Multireplacer>> {
  const rawRecords = await parseCSV(buffer);

  const dtos: FlavorizationRuleDTO[] = rawRecords.map((r, rowIndex) => ({
    rowIndex,
    disabled: r.disabled || '',
    id: r.id || '',
    flavorizationLevel: r.flavorizationLevel || '',
    match: r.match || '',
    flags: r.flags || '',
    partOfSpeech: r.partOfSpeech || '',
    genesis: r.genesis || '',
    replacement1: r.replacement1 || '',
    replacement2: r.replacement2 || '',
    replacement3: r.replacement3 || '',
    replacement4: r.replacement4 || '',
    replacement5: r.replacement5 || '',
  }));

  const replacers: Record<keyof typeof FlavorizationLevel, Multireplacer> = {
    Reverse: new Multireplacer(),
    Standard: new Multireplacer(),
    Silly: new Multireplacer(),
    Etymological: new Multireplacer(),
  };

  for (const dto of dtos) {
    if (dto.disabled || !dto.match) {
      continue;
    }

    if (!dto.id) {
      throw new Error(`Rule at row ${dto.rowIndex} has no ID`);
    }

    const rule = new MultireplacerRule(dto.id);
    const flags = new Set(dto.flags.split(/\s+/));
    rule.searchValue = flags.has('FIXED')
      ? dto.match
      : new RegExp(dto.match, 'g');

    if (flags.has('BY-WORD')) {
      rule.splitter = 'word';
    } else if (flags.has('BY-LETTER')) {
      rule.splitter = 'letter';
    }

    if (dto.genesis) {
      let isApproximate = false;
      let isNegated = false;

      const genesisSet = new Set(
        dto.genesis
          .split('')
          .filter((letter) => {
            switch (letter) {
              case '?':
                isApproximate = true;
                return false;
              case '!':
                isNegated = true;
                return false;
              default:
                return true;
            }
          })
          .map((letter) => steenparse.genesis(letter)),
      );

      const predicate: FlavorizationPredicate<InterslavicIntermediate> =
        isApproximate
          ? (p) => !p.context.genesis || genesisSet.has(p.context.genesis)
          : (p) => !!p.context.genesis && genesisSet.has(p.context.genesis);

      if (isNegated) {
        rule.predicates.andNot(predicate);
      } else {
        rule.predicates.and(predicate);
      }
    }

    if (dto.partOfSpeech && !dto.partOfSpeech.startsWith('*')) {
      const isNegated = dto.partOfSpeech.startsWith('!');
      const partOfSpeech = _.pickBy(
        steenparse.partOfSpeech(dto.partOfSpeech.slice(isNegated ? 1 : 0)),
        _.identity,
      );

      const predicate: FlavorizationPredicate<InterslavicIntermediate> = (
        p,
      ) => {
        return _.isMatch(p.context.partOfSpeech, partOfSpeech);
      };

      if (isNegated) {
        rule.predicates.andNot(predicate);
      } else {
        rule.predicates.and(predicate);
      }
    }

    for (const replacement of [
      dto.replacement1,
      dto.replacement2,
      dto.replacement3,
      dto.replacement4,
      dto.replacement5,
    ]) {
      if (!replacement.trim()) {
        continue;
      }

      if (replacement === 'NULL') {
        rule.replacements.add('');
      } else if (flags.has('MAP')) {
        rule.replacements.add(createMapReplacer(replacement));
      } else {
        rule.replacements.add(replacement);
      }
    }

    switch (dto.flavorizationLevel) {
      case FlavorizationLevel.Silly:
        replacers.Silly.rules.add(rule);
        break;
      case FlavorizationLevel.Etymological:
        replacers.Etymological.rules.add(rule);
        break;
      case FlavorizationLevel.Reverse:
        replacers.Reverse.rules.add(rule);
        break;
      default:
        replacers.Silly.rules.add(rule);
        replacers.Standard.rules.add(rule);
        replacers.Etymological.rules.add(rule);
        break;
    }
  }

  return replacers;
}
