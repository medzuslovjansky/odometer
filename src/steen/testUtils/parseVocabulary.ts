import { BareRecord } from '../../types/BareRecord';
import { parse as steenparse } from '@interslavic/steen-utils';
import { parseCSV } from './parseCSV';

export async function parseVocabulary(
  isv: Buffer | string,
  lang: Buffer | string,
): Promise<BareRecord[]> {
  const rawIsv = await parseCSV(isv);
  const rawLang = await parseCSV(lang);

  const translations = new Map(rawLang.map((r) => [+r.id, r.translation]));

  return rawIsv.map<BareRecord>((r) => {
    const id = +r.id;
    const partOfSpeech = steenparse.partOfSpeech(r.partOfSpeech);
    const genesis = r.etymology ? steenparse.genesis(r.etymology) : undefined;
    const isPhrase = partOfSpeech.name === 'phrase';
    const isv = steenparse.synset(r.lemma, { isPhrase });
    const rawTranslation = translations.get(id);
    const translation = rawTranslation
      ? steenparse.synset(rawTranslation, {
          isPhrase,
        })
      : undefined;

    return {
      id,
      isv,
      translation,
      partOfSpeech,
      genesis,
    };
  });
}
