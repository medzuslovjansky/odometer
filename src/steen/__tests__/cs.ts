import { core } from '@interslavic/steen-utils';
import { promises as fs } from 'fs';
import { Odometer } from '../../core';
import { parseRuleSheet } from '../testUtils/parseRuleSheet';
import { parseVocabulary } from '../testUtils/parseVocabulary';

describe('Czech', () => {
  it('should be that', async () => {
    const [rulesBuffer, isvBuffer, translationBuffer] = await Promise.all(
      ['rules_cs', 'vocab_isv', 'translation_cs'].map((name) =>
        fs.readFile(require.resolve(`../../__fixtures__/${name}.csv`)),
      ),
    );

    const rules = await parseRuleSheet(rulesBuffer);
    const translations = await parseVocabulary(isvBuffer, translationBuffer);

    const values = (s: core.Synset) => [...s.lemmas()].map((l) => l.value);
    const odometer = new Odometer();
    for (const t of translations) {
      if (!t.translation || t.id > 30) {
        continue;
      }

      const theLang = rules.Reverse.process(values(t.translation), t);
      const theISV = rules.Standard.process(values(t.isv), t);
      const { a, b, distance } = odometer.getDifference(
        theISV.variants,
        theLang.variants,
      );

      if (a && b) {
        const chainA = [...a.replacements()]
          .map((r) => rules.Standard.rules.find(r))
          .map((r) => `${r && r.id}`)
          .join(', ');

        const chainB = [...b.replacements()]
          .map((r) => rules.Standard.rules.find(r))
          .map((r) => `${r && r.id}`)
          .join(', ');

        let report = '';
        report += `(${t.isv}) ∩ (${t.translation}) = ${a.root} ÷ ${b.root}`;
        report += `\nISV: ${a.root} → ${a} (${chainA})`;
        report += `\nCS: ${b.root} → ${b} (${chainB})`;
        report += `\nΔ(${a}, ${b}) ≈ ${distance}`;

        expect(report).toMatchSnapshot(`${t.id}: ${t.isv}`);
      }
    }
  });
});
