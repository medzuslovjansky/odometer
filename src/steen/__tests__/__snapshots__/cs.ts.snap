// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Czech should be that: 1: oko 1`] = `
"(oko) ∩ (oko) = oko ÷ oko
ISV: oko → oko ()
CS: oko → oko ()
Δ(oko, oko) ≈ 0"
`;

exports[`Czech should be that: 2: opisati 1`] = `
"(opisati) ∩ (popsat) = opisati ÷ popsat
ISV: opisati → opisat (Ko2)
CS: popsat → popsat ()
Δ(opisat, popsat) ≈ 0.3333333333333333"
`;

exports[`Czech should be that: 3: opera 1`] = `
"(opera) ∩ (opera) = opera ÷ opera
ISV: opera → opera ()
CS: opera → opera ()
Δ(opera, opera) ≈ 0"
`;

exports[`Czech should be that: 4: nikto 1`] = `
"(nikto) ∩ (nikdo) = nikto ÷ nikdo
ISV: nikto → nikto ()
CS: nikdo → nikdo ()
Δ(nikto, nikdo) ≈ 0.2"
`;

exports[`Czech should be that: 5: tęžky 1`] = `
"(tęžky) ∩ (těžký, silný, hustý) = tęžky ÷ těžký
ISV: tęžky → tježký (A2, B4)
CS: těžký → těžký ()
Δ(tježký, těžký) ≈ 0.2727272727272727"
`;

exports[`Czech should be that: 6: veliky 1`] = `
"(veliky) ∩ (velký) = veliky ÷ velký
ISV: veliky → veliký (A2)
CS: velký → velký ()
Δ(veliký, velký) ≈ 0.18181818181818182"
`;

exports[`Czech should be that: 7: osųđati 1`] = `
"(osųđati) ∩ (odsuzovat) = osųđati ÷ odsuzovat
ISV: osųđati → osuzat (B14, B3, Ko2)
CS: odsuzovat → odsuzovat ()
Δ(osuzat, odsuzovat) ≈ 0.4"
`;

exports[`Czech should be that: 8: odinųd 1`] = `
"(odinųd) ∩ (odjinud) = odinųd ÷ odjinud
ISV: odinųd → odinud (B14)
CS: odjinud → odjinud ()
Δ(odinud, odjinud) ≈ 0.15384615384615385"
`;

exports[`Czech should be that: 10: olovo 1`] = `
"(olovo) ∩ (olovo) = olovo ÷ olovo
ISV: olovo → olovo ()
CS: olovo → olovo ()
Δ(olovo, olovo) ≈ 0"
`;

exports[`Czech should be that: 11: abdikovati 1`] = `
"(abdikovati) ∩ (abdikovat) = abdikovati ÷ abdikovat
ISV: abdikovati → abdikovat (Ko2)
CS: abdikovat → abdikovat ()
Δ(abdikovat, abdikovat) ≈ 0"
`;

exports[`Czech should be that: 12: trava 1`] = `
"(trava) ∩ (tráva) = trava ÷ tráva
ISV: trava → trava ()
CS: tráva → tráva ()
Δ(trava, tráva) ≈ 0.1"
`;

exports[`Czech should be that: 13: osnovyvati 1`] = `
"(osnovyvati) ∩ (zakládat, ustanovovat, stanovit) = osnovyvati ÷ ustanovovat
ISV: osnovyvati → osnovyvat (Ko2)
CS: ustanovovat → ustanovovat ()
Δ(osnovyvat, ustanovovat) ≈ 0.4"
`;

exports[`Czech should be that: 14: ocěniti 1`] = `
"(ocěniti) ∩ (ocenit, ohodnotit) = ocěniti ÷ ocenit
ISV: ocěniti → ocenit (Dě, Ko2)
CS: ocenit → ocenit ()
Δ(ocenit, ocenit) ≈ 0"
`;

exports[`Czech should be that: 15: absencija 1`] = `
"(absencija) ∩ (absence, nepřítomnost) = absencija ÷ absence
ISV: absencija → absence (I5)
CS: absence → absence ()
Δ(absence, absence) ≈ 0"
`;

exports[`Czech should be that: 16: želųdòk 1`] = `
"(želųdòk) ∩ (žaludek) = želųdòk ÷ žaludek
ISV: želųdòk → želudòk (B14)
CS: žaludek → žaludek ()
Δ(želudòk, žaludek) ≈ 0.2857142857142857"
`;

exports[`Czech should be that: 17: råzměniti 1`] = `
"(råzměniti) ∩ (rozměnit) = råzměniti ÷ rozměnit
ISV: råzměniti → rozměnit (Ko2, P2)
CS: rozměnit → rozměnit ()
Δ(rozměnit, rozměnit) ≈ 0"
`;

exports[`Czech should be that: 18: astronomija 1`] = `
"(astronomija) ∩ (astronomie) = astronomija ÷ astronomie
ISV: astronomija → astronomie (I5)
CS: astronomie → astronomie ()
Δ(astronomie, astronomie) ≈ 0"
`;

exports[`Czech should be that: 20: dostavjati 1`] = `
"(dostavjati) ∩ (poskytovat) = dostavjati ÷ poskytovat
ISV: dostavjati → dostavět (K12, Ko2)
CS: poskytovat → poskytovat ()
Δ(dostavět, poskytovat) ≈ 0.5555555555555556"
`;

exports[`Czech should be that: 22: tŀpa 1`] = `
"(tŀpa) ∩ (dav, zástup, nával, tlačenice) = tŀpa ÷ nával
ISV: tŀpa → tŀpa ()
CS: nával → nával ()
Δ(tŀpa, nával) ≈ 0.8888888888888888"
`;

exports[`Czech should be that: 24: zapŀnjati 1`] = `
"(zapŀnjati) ∩ (vyplňovat) = zapŀnjati ÷ vyplňovat
ISV: zapŀnjati → zapŀňat (D7, Ko2)
CS: vyplňovat → vyplňovat ()
Δ(zapŀňat, vyplňovat) ≈ 0.625"
`;

exports[`Czech should be that: 25: zaregistrovati 1`] = `
"(zaregistrovati) ∩ (zaregistrovat) = zaregistrovati ÷ zaregistrovat
ISV: zaregistrovati → zaregistrovat (Ko2)
CS: zaregistrovat → zaregistrovat ()
Δ(zaregistrovat, zaregistrovat) ≈ 0"
`;

exports[`Czech should be that: 26: rotacija 1`] = `
"(rotacija) ∩ (rotace) = rotacija ÷ rotace
ISV: rotacija → rotace (I5)
CS: rotace → rotace ()
Δ(rotace, rotace) ≈ 0"
`;

exports[`Czech should be that: 27: råzličıje 1`] = `
"(råzličıje) ∩ (rozdíl) = råzličıje ÷ rozdíl
ISV: råzličıje → rozličıje (P2)
CS: rozdíl → rozdíl ()
Δ(rozličıje, rozdíl) ≈ 0.7333333333333333"
`;

exports[`Czech should be that: 28: žiti 1`] = `
"(žiti) ∩ (žít, být živý) = žiti ÷ žít
ISV: žiti → žit (Ko2)
CS: žít → žít ()
Δ(žit, žít) ≈ 0.16666666666666666"
`;

exports[`Czech should be that: 29: časovy 1`] = `
"(časovy) ∩ (dočasný) = časovy ÷ dočasný
ISV: časovy → časový (A2)
CS: dočasný → dočasný ()
Δ(časový, dočasný) ≈ 0.6153846153846154"
`;

exports[`Czech should be that: 30: krajno 1`] = `
"(krajno) ∩ (krajně, velmi) = krajno ÷ krajně
ISV: krajno → krajno ()
CS: krajně → krajně ()
Δ(krajno, krajně) ≈ 0.16666666666666666"
`;
